pkg/token/identitytoken/identitytoken.go

```diff
@@ -0,0 +1,156 @@
+        Expiry          time.Time `json:"expiry"`
+        ExpiresIn       int       `json:"expires_in"`
+        Scope           string    `json:"scope"`
+        AccessTokenOnly bool      `json:"accessTokenOnly"`
+}
+
+type Client struct {
+        identityServiceURL string
+        httpClient         httpClient
+}
+
+type httpClient interface {
+        Do(req *http.Request) (*http.Response, error)
+}
+
+// New creates a new identity Client object
+func New(identityServiceURL string) *Client {
+        client := &http.Client{Timeout: 10 * time.Second}
+        identityServiceURL = strings.TrimRight(identityServiceURL, "/")
+        return &Client{
+                identityServiceURL: identityServiceURL,
+                httpClient:         client,
+        }
+}
+
+func doRetries(call func() (*http.Response, error), retries int) (*http.Response, error) {
+        var resp *http.Response
+        var err error
+
+        for {
+                resp, err = call()
+                if err != nil {
+                        return nil, err
+                }
+
+                if !isStatusRetryable(resp.StatusCode) || retries == 0 {
+                        break
+                }
+
+                time.Sleep(3 * time.Second)
+                retries--
+        }
+
+        return resp, nil
+}
+
+func (c *Client) GenerateToken(ctx context.Context, tenantID, clientID, clientSecret string) (string, error) {
+        params := GenerateTokenInput{
+                TenantID:     tenantID,
+                ClientID:     clientID,
+                ClientSecret: clientSecret,
+                GrantType:    "client_credentials",
+        }
+
+        url := fmt.Sprintf("%s/v1/token", c.identityServiceURL)
+
+        b, err := json.Marshal(params)
+        if err != nil {
+                return "", err
+        }
+
+        req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, strings.NewReader(string(b)))
+
+        if err != nil {
+                return "", err
+        }
+        req.Header.Set("Content-Type", "application/json")
+
+        resp, err := doRetries(func() (*http.Response, error) {
+                return c.httpClient.Do(req)
+        }, retryLimit)
+        if err != nil {
+                return "", err
+        }
+        defer resp.Body.Close()
+
+        switch resp.StatusCode {
+        case http.StatusOK:
+                break
+        case http.StatusBadRequest:
+                body, err := ioutil.ReadAll(resp.Body)
+                if err != nil {
+                        return "", err
+                }
+                msg := fmt.Sprintf("Bad request: %v", string(body))
+                err = errors.MakeErrBadRequest(errors.ErrorResponse{
+                        ErrorCode: "ErrGenerateTokenBadRequest",
+                        Message:   msg,
+                })
+
+                return "", err
+        case http.StatusUnauthorized, http.StatusForbidden:
+                err = errors.MakeErrForbidden(clientID)
```

Comment:
I think callers of GenerateToken need to be able to distinguish between 401 and 403.
(It's not clear if that's the case here.) It would be good to distinguish between them and
update the unit tests to have one for the 401 case and one for the 403 case.

ie split into MakeErrForbidden/MakeErrUnauthorized

Reply:
Yes sure, I split it as you advice to make it more clear.